Safely close multiple resources in RapidsBufferCatalog

jihoonson opened the pull request:
Minor improvement to safely close all resources in `RapidsBufferCatalog` even if one of them fails to close.

Contributes to #10502

jlowe commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -864,30 +865,16 @@ object RapidsBufferCatalog extends Logging {
-      diskStorage = null
-    }
+    Seq(_singleton, memoryEventHandlerCloser, deviceStorage, hostStorage, diskStorage)
+      .filter(e => e != null).safeClose()
```
I don't think the filter is necessary since safeClose handles null arguments.

jihoonson replied:
Thanks, removed it.

gerashegalov commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -727,9 +729,8 @@ class RapidsBufferCatalog(
-    bufferIdToHandles.values.forEach { handles =>
-      handles.foreach(_.close())
-    }
+    bufferIdToHandles.values.asScala.toSeq.flatMap(_.seq).safeClose()
```
The existing code already creates a throwaway collection via `values`. This PR adds more conversions doing it on top of it.

```suggestion
    bufferIdToHandles.forEach((_, bufSeq) => bufSeq.safeClose())
```

jihoonson replied:
The thing is that safeClose() can throw. I think iterating in forEach will not complete the iteration if one of safeClose() calls throws.

jihoonson replied:
Also, `ConcurrentHashMap.values()` creates a view of it. It shouldn't be expensive to create.

gerashegalov replied:
but then we have 
- asScala 
- toSeq
- flatMap(_.seq)

gerashegalov replied:
> The thing is that safeClose()  can throw.

good point

jihoonson replied:
> but then we have
> 
> * asScala
> * toSeq
> * flatMap(_.seq)

Hmm, does this make a copy of the values collection? If so, I could use the Java Stream API with a try-catch clause.

jihoonson replied:
I thought it would make a scala `Iterable` per https://docs.scala-lang.org/overviews/collections-2.13/conversions-between-java-and-scala-collections.html as `values` returns a java Collection.

gerashegalov replied:
Yes. Stream API will also make a copy. Not worth it, I thought there is a simple solution. If we had more nested use cases we could add another implicit

jihoonson replied:
> Stream API will also make a copy.

I would be surprised if it does. My understanding is that they are lazily evaluated and never be materialized unless you tell it to do so. It is just a fancier version of the `Iterator` API.

gerashegalov replied:
Checking in REPL there are only two materializations in the original 

`values.asScala.toSeq.flatMap(_.seq)`

from toSeq  retrurning a List and  consequently from the following `flatMap`

This  `bufferIdToHandles.values.asScala.flatten.safeClose()` saves toSeq but we need to fix our scala-2.12 implictis.


```scala
scala> val cm = new java.util.concurrent.ConcurrentHashMap[String,Seq[String]]()
val cm: java.util.concurrent.ConcurrentHashMap[String,Seq[String]] = {}

scala> cm.put("test_key", Seq("test_value_1", "test_value_2"))
val res11: Seq[String] = null

scala> cm.values
val res12: java.util.Collection[Seq[String]] = [List(test_value_1, test_value_2)]

scala> cm.values.asScala
warning: 1 deprecation (since 2.13.0); for details, enable `:setting -deprecation` or `:replay -deprecation`
val res13: Iterable[Seq[String]] = Iterable(List(test_value_1, test_value_2))

scala> cm.values.asScala.toSeq
warning: 1 deprecation (since 2.13.0); for details, enable `:setting -deprecation` or `:replay -deprecation`
val res14: Seq[Seq[String]] = List(List(test_value_1, test_value_2))

scala> cm.values.asScala.toSeq.flatMap(_.seq)
warning: 2 deprecations (since 2.13.0); for details, enable `:setting -deprecation` or `:replay -deprecation`
val res15: Seq[String] = List(test_value_1, test_value_2)
```

jihoonson replied:
Surprising that `flatMap` materializes.. 🙁 Good to know. Yeah perhaps it's a good idea to add a new `safeClose()` version for nested collections if we have enough use cases. It could use `bufferIdToHandles.values.asScala.flatten.safeClose()` you suggested.

gerashegalov replied:
> Surprising that flatMap materializes..

The culprit is toSeq, which you only have to do because Scala 2.12 version of implicitcs is written for SeqLike. You can update it in this PR to use Iterable as in https://github.com/NVIDIA/spark-rapids/blob/8ef2e41d4b4126eecf9130043a2b7bedf2f4263f/sql-plugin/src/main/scala-2.13/com/nvidia/spark/rapids/implicits.scala#L86C2-L112C4

jihoonson replied:
Thanks for the suggestion. I think it would be better to be done in a separate PR though, created https://github.com/NVIDIA/spark-rapids/issues/11314.

gerashegalov commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -864,30 +865,15 @@ object RapidsBufferCatalog extends Logging {
-      //Rmm.clearEventHandler()
-      memoryEventHandler = null
-    }
+    val memoryEventHandlerCloser: AutoCloseable = () =>
+      if (memoryEventHandler != null) {
+        // Workaround for shutdown ordering problems where device buffers allocated
+        // with this handler are being freed after the handler is destroyed
+        //Rmm.clearEventHandler()
+        memoryEventHandler = null
+      }
```
nit: does not safe lines of code and creates an additional class for lambda, maybe just leave it as is?

jihoonson replied:
Same here. These resources should be closed even if one of them fails to close. The use of lambda was added to keep the original order of clearing resources.

jihoonson commented:
Build

jihoonson commented:
Can someone help to run the CI? I don't seem to have enough power for that.

gerashegalov commented:
build

jihoonson commented:
Looking into the CI failure. Seems like a legit issue.

jihoonson commented:
> Looking into the CI failure. Seems like a legit issue.

It was because of the missing nullifying. Fixed in https://github.com/NVIDIA/spark-rapids/pull/11307/commits/6607bbb3f87bc3e0b18ec7ea3684621d9d3942ed. As part of this, I removed [the `AutoCloseable`  workaround that only sets `memoryEventHandler` to null](https://github.com/NVIDIA/spark-rapids/pull/11307#discussion_r1710135240), hoping that this would have no impact. It is a new behavior that `RapidsBufferCatalog` closes all its resources and then nullifies them. The previous behavior was to close and nullify them one by one.

jihoonson commented:
build

gerashegalov reviewed (approved):
LGTM
